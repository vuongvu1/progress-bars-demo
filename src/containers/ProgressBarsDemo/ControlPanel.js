import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ProgressBar from 'components/ProgressBar';

// #region Styles
const Wrapper = styled.div`
  width: 400px;
  background-color: ${(props) => props.theme.lightBlue};
  padding: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
`;

const Title = styled.div`
  font-size: ${(props) => props.theme.fontSizeBig};
  margin-bottom: 20px;
`;

const BarsWrapper = styled.div`
  width: 100%;

  * {
    margin-bottom: 10px;
  }
`;

const ControlPanelWrapper = styled.div`
  width: 100%;
  text-align: center;

  * {
    margin-right: 5px;
  }
`;
// #endregion Styles


const ControlPanel = ({ barList, buttonList, limitValue }) => {
  const [activeBar, setActiveBar] = useState('');
  const [bars, setBars] = useState(barList);
  const [buttons, setButtons] = useState(buttonList);

  useEffect(() => { setBars(barList); }, [barList]);
  useEffect(() => { setButtons(buttonList); }, [buttonList]);

  const changeValue = useCallback((value) => {
    const newBars = bars.map((bar) => {
      if (activeBar === bar.id) {
        const newValue = bar.data + value;
        return {
          ...bar,
          data: newValue < 0 ? 0 : newValue,
        };
      }
      return bar;
    });
    setBars(newBars);
  }, [activeBar, bars]);

  return (
    <Wrapper>
      <Title>
        {activeBar
          ? `Selected (${activeBar})`
          : 'Select a bar to update'}
      </Title>
      <BarsWrapper>
        {bars.map(({ id, data }) => (
          <ProgressBar
            key={id}
            id={id}
            value={data}
            limit={limitValue}
            isActive={activeBar === id}
            setActive={setActiveBar}
          />
        ))}
      </BarsWrapper>

      <ControlPanelWrapper>
        {buttons.map(({ id, data }) => (
          <input
            key={id}
            type='button'
            value={data > 0 ? `+${data}` : data}
            onClick={() => changeValue(data)}
          />
        ))}
      </ControlPanelWrapper>
    </Wrapper>
  );
};

ControlPanel.defaultProps = {
  barList: [],
  buttonList: [],
  limitValue: 0,
};

ControlPanel.propTypes = {
  barList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    data: PropTypes.number,
  })),
  buttonList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    data: PropTypes.number,
  })),
  limitValue: PropTypes.number,
};

export default ControlPanel;
