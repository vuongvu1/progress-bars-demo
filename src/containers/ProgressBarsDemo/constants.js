
export const actionTypes = {
  FETCH_BARS_DATA: 'FETCH_BARS_DATA',
  FETCH_BARS_DATA_SUCCESS: 'FETCH_BARS_DATA_SUCCESS',
  FETCH_BARS_DATA_FAILURE: 'FETCH_BARS_DATA_FAILURE',
};

export default actionTypes;
