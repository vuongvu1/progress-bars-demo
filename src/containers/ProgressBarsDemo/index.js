import React from 'react';

const ProgressBarsDemo = () => {

  return <div>This is ProgressBarsDemo</div>;
};

export default ProgressBarsDemo;
