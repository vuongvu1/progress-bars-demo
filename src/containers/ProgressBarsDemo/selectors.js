// import { createSelector } from 'reselect';

export const barListSelector = (state) => state.progressBar.bars.list;
export const buttonListSelector = (state) => state.progressBar.buttons.list;

export const allBarsSelector = (state) => state.progressBar.bars.all;
export const allButtonSelector = (state) => state.progressBar.buttons.all;

export const limitValueSelector = (state) => state.progressBar.limit;
export const loadingSelector = (state) => state.progressBar.isLoading;
