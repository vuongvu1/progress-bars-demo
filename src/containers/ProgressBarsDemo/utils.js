
export const normalizeBarsData = ({ bars, buttons, limit }) => {
  const barsObject = bars.reduce((object, bar, index) => ({
    ...object,
    [`bar_id_${index}`]: bar,
  }), {});

  const buttonsObject = buttons.reduce((object, button, index) => ({
    ...object,
    [`button_id_${index}`]: button,
  }), {});

  return {
    bars: {
      ...barsObject,
      allBars: bars,
    },
    buttons: {
      ...buttonsObject,
      allButtons: buttons,
    },
    limit,
  };
};

export default normalizeBarsData;
