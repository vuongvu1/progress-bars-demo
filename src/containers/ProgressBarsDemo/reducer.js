
import { actionTypes } from './constants';

export const initialState = {
  isLoading: false,
  bars: [],
  buttons: [],
  limit: 0,
};

const progressBarsReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.FETCH_BARS_DATA: {
      return {
        ...state,
        isLoading: true,
      };
    }

    case actionTypes.FETCH_BARS_DATA_SUCCESS: {
      const { bars, buttons, limit } = action.payload;

      return {
        ...state,
        isLoading: false,
        bars,
        buttons,
        limit,
      };
    }

    case actionTypes.FETCH_BARS_DATA_FAILURE: {
      return {
        ...state,
        isLoading: false,
      };
    }

    default:
      return state;
  }
};

export default progressBarsReducer;
