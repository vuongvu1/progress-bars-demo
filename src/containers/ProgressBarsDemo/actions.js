import { fetchBarsData } from 'utils/apiClient';
import { actionTypes } from './constants';
import { normalizeBarsData } from './utils';

export const fetchBarsDataAction = () => async (dispatch) => {
  dispatch({ type: actionTypes.FETCH_BARS_DATA });

  try {
    const barsData = await fetchBarsData();

    dispatch({
      type: actionTypes.FETCH_BARS_DATA_SUCCESS,
      payload: normalizeBarsData(barsData),
    });
  } catch (err) {
    dispatch({ type: actionTypes.FETCH_BARS_DATA_FAILURE });
  }
};

export default fetchBarsDataAction;
