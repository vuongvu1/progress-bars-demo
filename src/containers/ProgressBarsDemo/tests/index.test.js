import React from 'react';
import { shallow } from 'enzyme';
import ProgressBarsDemo from '../index';

describe('<ProgressBarsDemo />', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<ProgressBarsDemo />);
    expect(wrapper).toMatchSnapshot();
  });
});
