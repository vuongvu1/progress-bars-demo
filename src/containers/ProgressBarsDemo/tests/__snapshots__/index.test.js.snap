// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBarsDemo /> renders without crashing 1`] = `ShallowWrapper {}`;
